system_tests/das_test.go

```diff
@@ -0,0 +1,151 @@
+	"encoding/json"
+	"math/big"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/offchainlabs/nitro/blsSignatures"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/offchainlabs/nitro/arbnode"
+
+	"github.com/offchainlabs/nitro/das/dasrpc"
+
+	"github.com/offchainlabs/nitro/das"
+)
+
+func startLocalDASServer(t *testing.T, ctx context.Context, dataDir string) (*dasrpc.DASRPCServer, *blsSignatures.PublicKey, dasrpc.BackendConfig) {
+	lis, err := net.Listen("tcp", "localhost:0")
+	Require(t, err)
+	keyDir := t.TempDir()
+	pubkey, _, err := das.GenerateAndStoreKeys(keyDir)
+	Require(t, err)
+	dasConfig := das.LocalDiskDASConfig{
+		KeyDir:             keyDir,
+		DataDir:            dataDir,
+		StoreSignerAddress: "none",
+	}
+	localDas, err := das.NewLocalDiskDAS(dasConfig)
+	Require(t, err)
+	dasServer, err := dasrpc.StartDASRPCServerOnListener(ctx, lis, localDas)
+	Require(t, err)
+	config := dasrpc.BackendConfig{
+		URL:                 lis.Addr().String(),
+		PubKeyBase64Encoded: blsPubToBase64(pubkey),
+		SignerMask:          1,
+	}
+	return dasServer, pubkey, config
+}
+
+func blsPubToBase64(pubkey *blsSignatures.PublicKey) string {
+	pubkeyBytes := blsSignatures.PublicKeyToBytes(*pubkey)
+	encodedPubkey := make([]byte, base64.StdEncoding.EncodedLen(len(pubkeyBytes)))
+	base64.StdEncoding.Encode(encodedPubkey, pubkeyBytes)
+	return string(encodedPubkey)
+}
+
+func aggConfigForBackend(t *testing.T, backendConfig dasrpc.BackendConfig) das.AggregatorConfig {
+	backendsJsonByte, err := json.Marshal([]dasrpc.BackendConfig{backendConfig})
+	Require(t, err)
+	return das.AggregatorConfig{
+		AssumedHonest:      1,
+		Backends:           string(backendsJsonByte),
+		StoreSignerAddress: "none",
+	}
+}
+
+func TestDASRekey(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Setup L1 chain and contracts
+	chainConfig := params.ArbitrumDevTestDASChainConfig()
+	l1info, l1client, _, l1stack := CreateTestL1BlockChain(t, nil)
+	defer l1stack.Close()
+	addresses := DeployOnTestL1(t, ctx, l1info, l1client, chainConfig.ChainID)
+
+	// Setup DAS servers
+	dasDataDir := t.TempDir()
+	dasServerA, pubkeyA, backendConfigA := startLocalDASServer(t, ctx, dasDataDir)
+	authorizeDASKeyset(t, ctx, pubkeyA, l1info, l1client)
+
+	// Setup L2 chain
+	l2info, l2stack, l2chainDb, l2blockchain := createL2BlockChain(t, nil, chainConfig)
+
+	// Setup DAS config
+	l1NodeConfigA := arbnode.ConfigDefaultL1Test()
+	l1NodeConfigA.DataAvailability.ModeImpl = "aggregator"
+	l1NodeConfigA.DataAvailability.AggregatorConfig = aggConfigForBackend(t, backendConfigA)
+
+	sequencerTxOpts := l1info.GetDefaultTransactOpts("Sequencer", ctx)
+	sequencerTxOptsPtr := &sequencerTxOpts
+	nodeA, err := arbnode.CreateNode(l2stack, l2chainDb, l1NodeConfigA, l2blockchain, l1client, addresses, sequencerTxOptsPtr, nil)
+	Require(t, err)
+	Require(t, nodeA.Start(ctx))
+
+	nodeA.StopAndWait()
+
+	dasServerA.Stop()
```

Comment:
Do we need to send any transactions with the first node so that a batch gets posted with the first keyset?

Reply:
I think this had been working before, but I've now added a check to guarantee that we've posted a batch using the first DAS